src/SharedStream.mo

```diff
@@ -0,0 +1,380 @@
+import Debug "mo:base/Debug";
+import Error "mo:base/Error";
+import Int "mo:base/Int";
+import Iter "mo:base/Iter";
+import Nat "mo:base/Nat";
+import Option "mo:base/Option";
+import Principal "mo:base/Principal";
+import R "mo:base/Result";
+import Time "mo:base/Time";
+import Timer "mo:base/Timer";
+
+import Vec "mo:vector";
+import QueueBuffer "QueueBuffer";
+
+module {
+
+  type StreamError = { #NotRegistered; #StreamClosed };
+
+  /// Usage:
+  ///
+  /// let receiver = StreamReceiver<Int>(
+  ///   123
+  ///   func (streamId : Nat, element: Int, index: Nat): () {
+  ///     ... do your logic with incoming item
+  ///   }
+  /// );
+  ///
+  /// Hook-up receive function in the actor class:
+  /// public shared func onStreamChunk(streamId : Nat, chunk: [Int], firstIndex: Nat) : async () {
+  ///   switch (streamId) case (123) { await receiver.onChunk(chunk, firstIndex); }; case (_) { Error.reject("Unknown stream"); }; };
+  /// };
+  public class StreamReceiver<T>(
+    streamId : Nat,
+    callback : (streamId : Nat, item : T, index : Nat) -> (),
+    startFromIndex : Nat,
+  ) {
+
+    var expectedNextIndex_ : Nat = startFromIndex;
+
+    var lastChunkTimestamp : Time.Time = Time.now();
+
+    public func isStreamClosed() : Bool = (Time.now() - lastChunkTimestamp) > 120_000_000_000; // 2 minutes
+
+    public func onChunk(chunk : [T], firstIndex : Nat) : R.Result<(), { #StreamClosed }> {
+      if (isStreamClosed()) {
+        return #err(#StreamClosed);
+      };
+      if (firstIndex != expectedNextIndex_) {
+        Debug.trap("Broken chunk index: " # Nat.toText(firstIndex) # "; expected: " # Nat.toText(expectedNextIndex_));
+      };
+      for (index in chunk.keys()) {
+        callback(streamId, chunk[index], firstIndex + index);
+      };
+      expectedNextIndex_ += chunk.size();
+      #ok();
+    };
+
+  };
+
+  /// Usage:
+  ///
+  /// let sender = StreamSender<Int>(
+  ///   123,
+  ///   10,
+  ///   10,
+  ///   func (item) = 1,
+  ///   anotherCanister.appendStream,
+  /// );
+  /// sender.next(1);
+  /// sender.next(2);
+  /// .....
+  /// sender.next(12);
+  /// await* sender.sendChunk(); // will send (123, [1..10], 0) to `anotherCanister`
+  /// await* sender.sendChunk(); // will send (123, [11..12], 10) to `anotherCanister`
+  /// await* sender.sendChunk(); // will do nothing, stream clean
+  public class StreamSender<T>(
+    streamId : Nat,
+    maxSize : ?Nat,
+    weightLimit : Nat,
+    weightFunc : (item : T) -> Nat,
+    callback : (streamId : Nat, items : [T], firstIndex : Nat) -> async R.Result<(), StreamError>,
+  ) {
+
+    let queue : QueueBuffer.QueueBuffer<T> = QueueBuffer.QueueBuffer<T>();
+    // a head of queue before submitting lately failed chunk. Used for error-handling
+    var lowestError : { #Inf; #Val : Nat } = #Inf;
```

Comment:
Do we need the variant or can we just use type `?Nat`?

Reply:
switched to ?Nat